File: internal/preflight/preflight.go
// Package preflight validates external runtime dependencies before the bot
// starts provider schedulers.
package preflight

import (
	"fmt"
	"os"
	"os/exec"
	"sort"
	"strings"

	"tiket-kereta-notifier/internal/config"
)

// Check reports every runtime dependency missing for the enabled features.
// It never installs or changes software on the host.
func Check(cfg *config.Config) error {
	var missing []string
	requiredProviders := make(map[string]bool)
	for _, train := range cfg.FlatTrains {
		requiredProviders[strings.ToLower(train.ProviderName)] = true
	}

	if requiredProviders["bookingkai"] && !chromiumAvailable(cfg.Browser.ChromiumPath) {
		missing = append(missing, "chromium (run ./scripts/setup-ubuntu.sh or set browser.chromium_path)")
	}
	if requiredProviders["tiketcom"] && !commandAvailable("curl_chrome110") {
		missing = append(missing, "curl_chrome110 (run ./scripts/setup-ubuntu.sh)")
	}
	if cfg.Webhook.Enabled && !commandAvailable("cloudflared") {
		missing = append(missing, "cloudflared (run ./scripts/setup-ubuntu.sh)")
	}

	if len(missing) == 0 {
		return nil
	}
	sort.Strings(missing)
	return fmt.Errorf("missing runtime dependencies:\n- %s", strings.Join(missing, "\n- "))
}

func commandAvailable(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}

func chromiumAvailable(configuredPath string) bool {
	if configuredPath != "" {
		info, err := os.Stat(configuredPath)
		return err == nil && !info.IsDir() && info.Mode().Perm()&0o111 != 0
	}

	for _, name := range []string{"chromium-browser", "chromium", "google-chrome-stable", "google-chrome"} {
		if commandAvailable(name) {
			return true
		}
	}
	return false
}

File: notifier/preflight.py
"""Validates external runtime dependencies before the bot starts provider schedulers.

Never installs or changes anything on the host — only reports what's missing.
"""

from __future__ import annotations

import shutil

from .config import Config


class PreflightError(Exception):
    pass


def check(cfg: Config) -> None:
    missing: list[str] = []

    if cfg.webhook.enabled and not _command_available("cloudflared"):
        missing.append("cloudflared (run ./scripts/setup-ubuntu.sh)")

    if missing:
        details = "\n- ".join(sorted(missing))
        raise PreflightError(f"missing runtime dependencies:\n- {details}")


def _command_available(name: str) -> bool:
    return shutil.which(name) is not None
